## **EX NO:**
##   **DATE :**
# <center>**CROSS VALIDATION**</center>

## **AIM :**

### To understand and implement  cross-validation using python code in Linear Regression to evaluate model  performance and prevent overfitting.

#### **Importing Headers**

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split, cross_val_score,KFold
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

 #### **Load Dataset**

In [ ]:
import ssl
import certifi
from sklearn.datasets import fetch_california_housing

# Update SSL context to use certifi certificates
ssl._create_default_https_context = ssl._create_unverified_context

In [ ]:
from sklearn.datasets import fetch_california_housing
# Load dataset
data = fetch_california_housing()
df = pd.DataFrame(data.data, columns=data.feature_names)
df['Target'] = data.target
# Display first 5 rows
df.head()

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1  8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2  7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3  5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4  3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   

   Longitude  Target  
0    -122.23   4.526  
1    -122.22   3.585  
2    -122.24   3.521  
3    -122.25   3.413  
4    -122.25   3.422

### **Check for null values**

In [ ]:
df.isnull().sum()

MedInc        0
HouseAge      0
AveRooms      0
AveBedrms     0
Population    0
AveOccup      0
Latitude      0
Longitude     0
Target        0
dtype: int64

##### No null values so no need to handle

####  **Define Features and Target Variable**

BEFORE CROSS-VALIDATION

In [ ]:
from sklearn.metrics import  r2_score
# Define Features and Target Variable
X = df.drop(columns=['Target'])
y = df['Target']
# K-Fold Cross-Validation
# Define model
model = LinearRegression()
# Train-test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
# Train model
model.fit(X_train, y_train)
# Predict and evaluate
y_pred = model.predict(X_test)
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
print("Test Set RMSE:", rmse)
# evaluate the model on test set
r2 = r2_score(y_test, y_pred)
print("R square value:",r2)

Test Set RMSE: 0.7455813830127764
R square value: 0.5757877060324508


#### **Values before cross validation Test Set RMSE: 0.7455813830127764 , R square value: 0.5757877060324508**

**RMSE** VALUES FOR K-FOLD CROSS VALIDATION WHEN VALUES OF K=5

In [ ]:
# Define Features and Target Variable
X = df.drop(columns=['Target'])
y = df['Target']
# K-Fold Cross-Validation
# Define model
model = LinearRegression()
# Define K-Fold strategy
kf = KFold(n_splits=5, shuffle=True, random_state=42)
# Perform Cross-Validation
cv_scores = cross_val_score(model, X, y, cv=kf, scoring='neg_mean_squared_error')
# Convert negative MSE to positive RMSE
cv_rmse = np.sqrt(-cv_scores)
print("Cross-Validation RMSE Scores:", cv_rmse)
print("Mean RMSE:", np.mean(cv_rmse))

Cross-Validation RMSE Scores: [0.74558138 0.72639951 0.71364081 0.71052852 0.74510434]
Mean RMSE: 0.7282509142479742


**RMSE** VALUES FOR K-FOLD CROSS VALIDATION WHEN VALUES OF K=7

In [ ]:
# Define Features and Target Variable
X = df.drop(columns=['Target'])
y = df['Target']
# K-Fold Cross-Validation
# Define model
model = LinearRegression()
# Define K-Fold strategy
kf = KFold(n_splits=7, shuffle=True, random_state=42)
# Perform Cross-Validation
cv_scores = cross_val_score(model, X, y, cv=kf, scoring='neg_mean_squared_error')
# Convert negative MSE to positive RMSE
cv_rmse = np.sqrt(-cv_scores)
print("Cross-Validation RMSE Scores:", cv_rmse)
print("Mean RMSE:", np.mean(cv_rmse))

Cross-Validation RMSE Scores: [0.72976107 0.7235216  0.74812705 0.70706349 0.73549198 0.72382324
 0.73773824]
Mean RMSE: 0.7293609533531624


**RMSE** VALUES FOR K-FOLD CROSS VALIDATION WHEN VALUES OF K=9

In [ ]:
# Define Features and Target Variable
X = df.drop(columns=['Target'])
y = df['Target']
# K-Fold Cross-Validation
# Define model
model = LinearRegression()
# Define K-Fold strategy
kf = KFold(n_splits=9, shuffle=True, random_state=42)
# Perform Cross-Validation
cv_scores = cross_val_score(model, X, y, cv=kf, scoring='neg_mean_squared_error')
# Convert negative MSE to positive RMSE
cv_rmse = np.sqrt(-cv_scores)
print("Cross-Validation RMSE Scores:", cv_rmse)
print("Mean RMSE:", np.mean(cv_rmse))

Cross-Validation RMSE Scores: [0.74303916 0.73968962 0.71993089 0.72649736 0.7029506  0.72593626
 0.69929752 0.74695582 0.73970255]
Mean RMSE: 0.7271110855889659


# Summary 1:
**The RMSE values across different folds are quite stable, indicating consistent performance.
The test set RMSE (0.7456) is slightly higher than the cross-validation RMSE values, which might suggest minor overfitting.
Increasing K in cross-validation does not drastically change RMSE, which suggests the model is relatively stable.**

# EXERCISE:

#### **Importing Headers**

In [ ]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import KFold, GridSearchCV, LeaveOneOut, cross_val_score, train_test_split
from sklearn.metrics import mean_squared_error, r2_score, make_scorer

 #### **Load Dataset**

In [ ]:
# Load dataset
df = pd.read_csv("HousingData.csv")
df

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD  TAX  \
0    0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900    1  296   
1    0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671    2  242   
2    0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671    2  242   
3    0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622    3  222   
4    0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622    3  222   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...  ...   
501  0.06263   0.0  11.93   0.0  0.573  6.593  69.1  2.4786    1  273   
502  0.04527   0.0  11.93   0.0  0.573  6.120  76.7  2.2875    1  273   
503  0.06076   0.0  11.93   0.0  0.573  6.976  91.0  2.1675    1  273   
504  0.10959   0.0  11.93   0.0  0.573  6.794  89.3  2.3889    1  273   
505  0.04741   0.0  11.93   0.0  0.573  6.030   NaN  2.5050    1  273   

     PTRATIO       B  LSTAT  MEDV  
0       15.3  396.90   4.98  24.0  
1       17.8  396.90   9.14  21.6  
2       17.8  392.83   4.03  34.7  
3       18.7  394.63   2.94  33.4  
4       18.7  396.90    NaN  36.2  
..       ...     ...    ...   ...  
501     21.0  391.99    NaN  22.4  
502     21.0  396.90   9.08  20.6  
503     21.0  396.90   5.64  23.9  
504     21.0  393.45   6.48  22.0  
505     21.0  396.90   7.88  11.9  

[506 rows x 14 columns]

###**Handle missing values**

In [ ]:
df.isnull().sum()

CRIM       20
ZN         20
INDUS      20
CHAS       20
NOX         0
RM          0
AGE        20
DIS         0
RAD         0
TAX         0
PTRATIO     0
B           0
LSTAT      20
MEDV        0
dtype: int64

In [ ]:
# Replacing null values with mean or mode in columns without using inplace
df['CRIM'] = df['CRIM'].fillna(df['CRIM'].mean())
df['ZN'] = df['ZN'].fillna(df['ZN'].mean())
df['INDUS'] = df['INDUS'].fillna(df['INDUS'].mean())
df['LSTAT'] = df['LSTAT'].fillna(df['LSTAT'].mean())
df['CHAS'] = df['CHAS'].fillna(df['CHAS'].mode()[0])
df['AGE'] = df['AGE'].fillna(df['AGE'].mode()[0])


## **Define Features and Target Variable**

In [ ]:
# Define Features and Target Variable
X = df.drop(columns=['MEDV'])
y = df['MEDV']

## **1. Implement "Stratified K-Fold" Cross-Validation and compare results.**

In [ ]:
# K-Fold Cross-Validation
model = LinearRegression()
kf = KFold(n_splits=5, shuffle=True, random_state=42)

# Perform Cross-Validation
cv_scores = cross_val_score(model, X, y, cv=kf, scoring='neg_mean_squared_error')

# Convert negative MSE to positive RMSE
cv_rmse = np.sqrt(-cv_scores)
print("K-Fold Cross-Validation RMSE Scores:", cv_rmse)
print("Mean RMSE (K-Fold):", np.mean(cv_rmse))

K-Fold Cross-Validation RMSE Scores: [4.9906751  4.46962883 5.26537403 4.99625281 4.81385358]
Mean RMSE (K-Fold): 4.907156868906489


In [ ]:
# Stratified K-Fold Cross-Validation
# Bin the target variable (MEDV) into categories for stratification
y_binned = pd.qcut(y, q=9, labels=False)  # Binned into 9 quantiles

# Define the model
model = LinearRegression()

kf = KFold(n_splits=9, shuffle=True, random_state=42)
rmse_scores = []

# Perform Stratified K-Fold Cross-Validation
for train_index, test_index in kf.split(X, y_binned):  # Use binned target for stratification
    X_train, X_test = X.iloc[train_index], X.iloc[test_index]
    y_train, y_test = y.iloc[train_index], y.iloc[test_index]  # Use original target for regression

    # Fit the model
    model.fit(X_train, y_train)

    # Predict on the test set
    y_pred = model.predict(X_test)

    # Calculate RMSE for the fold
    rmse = np.sqrt(mean_squared_error(y_test, y_pred))
    rmse_scores.append(rmse)

# Output RMSE for each fold and the mean RMSE
print("Stratified K-Fold Cross-Validation RMSE Scores:", rmse_scores)
print("Mean RMSE (Stratified K-Fold):", np.mean(rmse_scores))

Stratified K-Fold Cross-Validation RMSE Scores: [3.688687004437897, 5.792164649739084, 4.206415816384389, 5.480981750383816, 5.846149554016179, 5.220092366109023, 4.996161388868331, 4.821243116673913, 4.817846763945882]
Mean RMSE (Stratified K-Fold): 4.985526934506502


In [ ]:
# Train-test split for final evaluation on the test set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Train model on the training set
model.fit(X_train, y_train)

# Predict on the test set
y_pred = model.predict(X_test)

# Calculate RMSE and R-squared for the test set
test_rmse = np.sqrt(mean_squared_error(y_test, y_pred))
test_r2 = r2_score(y_test, y_pred)
print("Test Set RMSE:", test_rmse)
print("R-squared on Test Set:", test_r2)

Test Set RMSE: 4.990675099725032
R-squared on Test Set: 0.6603633840004692


###**SUMMARY 1:**

#### The model performs is consistent for both K-Fold and Stratified K-Fold cross-validation, RMSE = 4.985.The test set RMSE=4.990 is very close to the cross-validation RMSE = 4.985 ,which suggests that the model does not suffer from significant overfitting or underfitting and model generalizes well to unseen data.

##**2. Experiment with different values of k in K-Fold Cross-Validation.**

In [ ]:
# Define Features and Target Variable
X = df.drop(columns=['MEDV'])
y = df['MEDV']

In [ ]:
# Define the model
model = LinearRegression()

In [ ]:
# Function to evaluate the model using K-Fold Cross-Validation for different values of k
def evaluate_k_fold(k_values):
    for k in k_values:
        print(f"\nEvaluating for K={k}")
        kf = KFold(n_splits=k, shuffle=True, random_state=42)

        # Perform Cross-Validation
        cv_scores = cross_val_score(model, X, y, cv=kf, scoring='neg_mean_squared_error')
        # Convert negative MSE to positive RMSE
        cv_rmse = np.sqrt(-cv_scores)
        print(f"Cross-Validation RMSE Scores: {cv_rmse}")
        print(f"Mean RMSE: {np.mean(cv_rmse)}")

# Experiment with different values of k
k_values = [3, 5, 7, 9, 10, 15]
evaluate_k_fold(k_values)


Evaluating for K=3
Cross-Validation RMSE Scores: [4.54877725 5.33096261 4.84847653]
Mean RMSE: 4.909405462315818

Evaluating for K=5
Cross-Validation RMSE Scores: [4.9906751  4.46962883 5.26537403 4.99625281 4.81385358]
Mean RMSE: 4.907156868906489

Evaluating for K=7
Cross-Validation RMSE Scores: [3.62596673 5.41645506 5.01238882 5.59467409 5.36051111 4.63097325
 4.85105519]
Mean RMSE: 4.927432033429858

Evaluating for K=9
Cross-Validation RMSE Scores: [3.688687   5.79216465 4.20641582 5.48098175 5.84614955 5.22009237
 4.99616139 4.82124312 4.81784676]
Mean RMSE: 4.985526934506502

Evaluating for K=10
Cross-Validation RMSE Scores: [3.84603434 5.91347794 4.05869189 4.85939351 6.14794584 4.39554627
 5.25606821 4.38345341 4.73298642 5.05107613]
Mean RMSE: 4.864467396723567

Evaluating for K=15
Cross-Validation RMSE Scores: [3.9746907  3.38544032 6.85741686 3.61826884 4.32476859 5.29344724
 7.11026648 3.80752223 4.70014938 5.19294763 5.30742785 4.00753235
 4.80967458 3.68397072 5.9257988

In [ ]:
# Final evaluation on the test set (for comparison)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

# Calculate RMSE and R-squared for the test set
test_rmse = np.sqrt(mean_squared_error(y_test, y_pred))
test_r2 = r2_score(y_test, y_pred)
print("Test Set RMSE:", test_rmse)
print("R-squared on Test Set:", test_r2)

Test Set RMSE: 4.990675099725032
R-squared on Test Set: 0.6603633840004692


###SUMMARY 2:

####When the values of K=15, offers lowest mean RMSE value but shows considerable variance across fold ,K=10 offers a balance between lower mean RMSE and acceptable variance, so K=10 is best choice for the model

##**3. Try "Leave-One-Out Cross-Validation (LOO-CV)" and analyze the computation time.**

In [ ]:
import time
# Measure computation time for Leave-One-Out Cross-Validation
#Leave-One-Out Cross-Validation
loo = LeaveOneOut()

start_time = time.time()  # Start time for LOO-CV

# Perform LOO-CV
cv_scores_loo = cross_val_score(model, X, y, cv=loo, scoring='neg_mean_squared_error')

# Convert negative MSE to positive RMSE
cv_rmse_loo = np.sqrt(-cv_scores_loo)
mean_rmse_loo = np.mean(cv_rmse_loo)

end_time = time.time()  # End time for LOO-CV
loo_time = end_time - start_time  # Compute time taken

print("LOO-CV RMSE Scores:", cv_rmse_loo[:5])  # Display only first 5 scores
print(f"Mean RMSE (LOO-CV): {mean_rmse_loo}")
print(f"LOO-CV Computation Time: {loo_time:.2f} seconds")

LOO-CV RMSE Scores: [ 6.35390281  3.52319911  4.09188796  4.55767592 11.55109995]
Mean RMSE (LOO-CV): 3.43251482211606
LOO-CV Computation Time: 1.78 seconds


In [ ]:
# Compare with K-Fold (k=5) Cross-Validation
kf = KFold(n_splits=5, shuffle=True, random_state=42)

start_time_kfold = time.time()  # Start time for K-Fold CV

# Perform K-Fold Cross-Validation
cv_scores_kfold = cross_val_score(model, X, y, cv=kf, scoring='neg_mean_squared_error')

# Convert negative MSE to positive RMSE
cv_rmse_kfold = np.sqrt(-cv_scores_kfold)
mean_rmse_kfold = np.mean(cv_rmse_kfold)

end_time_kfold = time.time()  # End time for K-Fold CV
kfold_time = end_time_kfold - start_time_kfold  # Compute time taken

print("K-Fold RMSE Scores:", cv_rmse_kfold)
print(f"Mean RMSE (K-Fold): {mean_rmse_kfold}")
print(f"K-Fold Computation Time: {kfold_time:.2f} seconds")

K-Fold RMSE Scores: [4.9906751  4.46962883 5.26537403 4.99625281 4.81385358]
Mean RMSE (K-Fold): 4.907156868906483
K-Fold Computation Time: 0.01 seconds


In [ ]:
# Final evaluation on the test set (for comparison)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

# Calculate RMSE and R-squared for the test set
test_rmse = np.sqrt(mean_squared_error(y_test, y_pred))
test_r2 = r2_score(y_test, y_pred)

print("Test Set RMSE:", test_rmse)
print("R-squared on Test Set:", test_r2)

Test Set RMSE: 4.990675099725032
R-squared on Test Set: 0.6603633840004692


##SUMMARY 3:
LOO-CV provides analysis of individual data points but computation time is significantly higher compared to K-Fold.
K-Fold Cross-Validation is more efficient and gives a reliable estimate of model performance.Hence k-fold CV is better choice for this model

###**4. Implement Grid Search Function with "GridSearchCV" for applying different values of K.**

Grid Search Cross-validation exhaustively tries every combination of the provided hyper-parameter values in order to find the best model.

In [ ]:
# Split the data into training and test sets (test set for final evaluation)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Define the model
model = LinearRegression()

# Define the parameter grid
param_grid = {
    'fit_intercept': [True, False],   # Whether to calculate the intercept
    'positive': [True, False]  # coefficients to be positive
}

# Define the K-Fold strategy within GridSearchCV
k_values = [3, 5, 7, 10, 15]  # Different values for the number of folds

# Perform grid search with cross-validation over different values of k
for k in k_values:
    print(f"\nGrid Search with K-Fold (k={k})")

    # Define K-Fold strategy
    kf = KFold(n_splits=k, shuffle=True, random_state=42)

    # Define the scoring metric as RMSE (Root Mean Squared Error)
    scorer = make_scorer(mean_squared_error, greater_is_better=False)

    # Set up GridSearchCV
    grid_search = GridSearchCV(estimator=model,
                               param_grid=param_grid,
                               scoring=scorer,
                               cv=kf,  # K-Fold Cross-Validation with k folds
                               refit=True,  # Refit the best model after search
                               n_jobs=-1)  # Use all available cores

    # Perform the grid search on the training data
    grid_search.fit(X_train, y_train)

    # Get the best model and best parameters
    best_model = grid_search.best_estimator_
    best_params = grid_search.best_params_

    print(f"Best Parameters for k={k}: {best_params}")

    # Evaluate the model on the training set using Cross-Validation RMSE
    best_cv_scores = cross_val_score(best_model, X_train, y_train, cv=kf, scoring='neg_mean_squared_error')
    best_cv_rmse = np.sqrt(-best_cv_scores)

    print(f"Best Cross-Validation RMSE Scores for k={k}: {best_cv_rmse}")
    print(f"Mean RMSE for k={k}: {np.mean(best_cv_rmse)}")


Grid Search with K-Fold (k=3)
Best Parameters for k=3: {'fit_intercept': True, 'positive': False}
Best Cross-Validation RMSE Scores for k=3: [5.0494285  5.55938925 4.37231263]
Mean RMSE for k=3: 4.993710126701333

Grid Search with K-Fold (k=5)
Best Parameters for k=5: {'fit_intercept': True, 'positive': False}
Best Cross-Validation RMSE Scores for k=5: [4.95181261 5.04074163 6.14366499 5.07813315 3.79347807]
Mean RMSE for k=5: 5.001566090141567

Grid Search with K-Fold (k=7)
Best Parameters for k=7: {'fit_intercept': True, 'positive': False}
Best Cross-Validation RMSE Scores for k=7: [5.48949742 4.3441783  5.31391041 6.29372717 4.33920472 5.00693845
 3.7107513 ]
Mean RMSE for k=7: 4.928315395939132

Grid Search with K-Fold (k=10)
Best Parameters for k=10: {'fit_intercept': True, 'positive': False}
Best Cross-Validation RMSE Scores for k=10: [3.7619451  5.94374992 5.29448449 4.61757017 6.53444837 5.53755384
 4.07823204 5.44085919 4.14157172 3.46521727]
Mean RMSE for k=10: 4.88156321189

In [ ]:
# Final model evaluation on test set
best_model.fit(X_train, y_train)
y_pred_test = best_model.predict(X_test)

# Calculate RMSE and R-squared for the test set
test_rmse = np.sqrt(mean_squared_error(y_test, y_pred_test))
test_r2 = r2_score(y_test, y_pred_test)

print("Test Set RMSE:", test_rmse)
print("R-squared on Test Set:", test_r2)

Test Set RMSE: 4.990675099725024
R-squared on Test Set: 0.6603633840004703


##SUMMARY 4:
The test-set RMSE is 4.9906 and by using Grid search with K-Fold K=3 provides a closer value.Therefore K=3 offers a balance between lower mean RMSE and acceptable variance, so K=3 is best choice for the model

In [3]:
from prettytable import PrettyTable

# Create table
table = PrettyTable()
table.field_names = ["Model", "Root Mean Squared Error (RMSE)"]

# Add rows
table.add_row([
    "Test Set :\n",
     4.99
])
table.add_row([
    "K-fold:\n",4.907])
table.add_row([
    "Stratified  K-fold:\n",
     4.985
])
table.add_row(["K-fold for K=3", 4.909])
table.add_row(["K-fold for K=10", 4.864])
table.add_row(["K-fold for K=15", 4.799])
# Set column alignments
table.align["Model"] = "l"
table.align["R Squared Value"] = "c"
table.align["Mean Squared Error (MSE)"] = "c"

# Print table
print(table)

+---------------------+--------------------------------+
| Model               | Root Mean Squared Error (RMSE) |
+---------------------+--------------------------------+
| Test Set :          |              4.99              |
|                     |                                |
| K-fold:             |             4.907              |
|                     |                                |
| Stratified  K-fold: |             4.985              |
|                     |                                |
| K-fold for K=3      |             4.909              |
| K-fold for K=10     |             4.864              |
| K-fold for K=15     |             4.799              |
+---------------------+--------------------------------+


In [6]:
from prettytable import PrettyTable

# Create table
table = PrettyTable()
table.field_names = ["Model", "Root Mean Squared Error (RMSE)","Computation Time"]

table.add_row([
    "K-fold:\n",4.907,0.01])
table.add_row([
    "LOO-CV:\n",
     3.432,1.78])

# Set column alignments
table.align["Model"] = "l"
table.align["Root Mean Squared Error (RMSE)"] = "c"
table.align["Computational Time"] = "c"

# Print table
print(table)

+---------+--------------------------------+------------------+
| Model   | Root Mean Squared Error (RMSE) | Computation Time |
+---------+--------------------------------+------------------+
| K-fold: |             4.907              |       0.01       |
|         |                                |                  |
| LOO-CV: |             3.432              |       1.78       |
|         |                                |                  |
+---------+--------------------------------+------------------+


#**RESULT:**

####Hence, the model was successfully executed, achieving a Test Set RMSE of 0.7456 with an R² value of 0.5758, indicating moderate accuracy. Thus, K-Fold Cross-Validation showed stable performance, with the lowest Mean RMSE of 0.7271 at K=9. Leave-One-Out Cross-Validation (LOO-CV) was not executed due to its high computational cost.